episodes/hpc.Rmd

```diff
@@ -0,0 +1,295 @@
+
+In the Parallel Processing section, we used the following configuration:
+```{R}
+library(crew)
+tar_option_set(
+  controller = crew_controller_local(workers = 2)
+)
+```
+To configure this for Slurm, we just swap out the controller with a new one from the `crew.cluster` package:
+
+```{R}
+library(crew.cluster)
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R"
+  )
+)
+```
+
+There are a number of options you can pass to `crew_controller_slurm()` to fine-tune the Slurm execution, [which you can find here](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html).
+Here we are only using two:
+
+  * `workers` sets the number of jobs that are submitted to Slurm to process targets.
+  * `script_lines` adds some lines to the Slurm submit script used by Targets. This is useful for loading Environment Modules and adding `#SBATCH` options.
+
+Let's run the modified workflow:
+
+```{R, eval=FALSE}
+source("R/packages.R")
+source("R/functions.R")
+
+library(crew.cluster)
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R"
+  )
+)
+
+tar_plan(
+  # Load raw data
+  tar_file_read(
+    penguins_data_raw,
+    path_to_file("penguins_raw.csv"),
+    read_csv(!!.x, show_col_types = FALSE)
+  ),
+  # Clean data
+  penguins_data = clean_penguin_data(penguins_data_raw),
+  # Build models
+  models = list(
+    combined_model = lm(
+      bill_depth_mm ~ bill_length_mm, data = penguins_data),
+    species_model = lm(
+      bill_depth_mm ~ bill_length_mm + species, data = penguins_data),
+    interaction_model = lm(
+      bill_depth_mm ~ bill_length_mm * species, data = penguins_data)
+  ),
+  # Get model summaries
+  tar_target(
+    model_summaries,
+    glance_with_mod_name_slow(models),
+    pattern = map(models)
+  ),
+  # Get model predictions
+  tar_target(
+    model_predictions,
+    augment_with_mod_name_slow(models),
+    pattern = map(models)
+  )
+)
+```
+
+::: challenge
+## Increasing Resources
+
+Q: How would you modify your `_targets.R` if your targets needed 200GB of RAM?
+
+::: hint
+Check the arguments for [`crew_controller_slurm`](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html#arguments-1).
+:::
+::: solution
+```R
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R",
+    # Added this
+    slurm_memory_gigabytes_per_cpu = 200,
+    slurm_cpus_per_task = 1
+  )
+)
+```
+:::
+:::
+
+## HPC Workers
+
+Despite what you might expect, `crew` does not submit one Slurm job for each target. 
+Instead, it uses persistent workers, meaning that you define a pool of workers when configuring the workflow.
+In our example above we used 3 workers.
+For each worker, `crew` submits a single Slurm job, and these workers will process multiple targets over their lifetime.
+
+We can verify that this has happened using `sacct`:
```

Comment:
To keep the material as agnostic to the scheduler as possible, I suggest removing this as learners not using Slurm would have no idea what `sacct` does and what the output means.

Reply:
The episode is already reasonably tied to Slurm by virtue of the controller class being Slurm-specific. Considering that, is there much benefit of removing this?

Reply:
```suggestion
We can verify that this has happened using `sacct`, which we can use to query information about our past jobs:
```

Yeah you're probably right that it is not that meaningful to remove it. But perhaps add a bit of an explainer to what the command is doing.

A semi-related question: Are there any monitoring tools available for the targets? I imagine that would also be useful for parallel execution.

Reply:
There are, but they're mostly covered already in the Lifecycle episode, or are extensions of the same functions: https://docs.ropensci.org/targets/reference/index.html#progress